### PR TITLE
✨ ⚠️ DeleteItemAction for volumesnapshots and volumesnapshotcontents

### PR DESCRIPTION
**Pull request by ashish-amarnath**
This change is incompatible with velero release `> v1.4 and <=v1.6`

Fixes #https://github.com/vmware-tanzu/velero/issues/3733

Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

### CODE REVIEW

**Comment by ashish-amarnath** [comment]
Testing
[asciicast](https://asciinema.org/a/410132)

[asciicast](https://asciinema.org/a/410131)

**Comment by ashish-amarnath** [comment]
Associated PR in the core velero repo https://github.com/vmware-tanzu/velero/pull/3734

**Review by nrb — COMMENTED** [review]
Quick feedback - can we add a compatibility matrix into the readme for this, so we don't miss it at release time?

**Comment by ashish-amarnath** [comment]
It did seem odd to refer to unreleased versions in the `README.md`. Nevertheless updated the Readme file.

File: internal/delete/volumesnapshotcontent_action.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	var snapCont snapshotv1beta1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), &snapCont); err != nil {
+		return errors.Wrapf(err, "failed to convert input.Item from unstructured")
+	}
+
+	// We don't want this DeleteItemAction plugin to delete VolumesnapshotContent taken outside of Velero.
+	// So skip deleting VolumesnapshotContent objects that were not created in the process of creating
+	// the Velero backup being deleted.
+	if !util.HasBackupLabel(&snapCont.ObjectMeta, input.Backup.Name) {
+		p.Log.Info("VolumesnapshotContent %s was not taken by backup %s, skipping deletion", snapCont.Name, input.Backup.Name)
```
Comment (dsu-igeek) [inline internal/delete/volumesnapshotcontent_action.go:43]: VolumesnapshotContent -> VolumeSnapshotContent

File: internal/delete/volumesnapshotcontent_action.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		return nil
+	}
+
+	p.Log.Infof("Deleting volumesnapshotcontent %s", snapCont.Name)
```
Comment (dsu-igeek) [inline internal/delete/volumesnapshotcontent_action.go:47]: VolumeSnapshotContent

File: internal/delete/volumesnapshotcontent_action.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	_, snapClient, err := util.GetClients()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	err = util.SetVolumeSnapshotContentDeletionPolicy(snapCont.Name, snapClient.SnapshotV1beta1())
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, fmt.Sprintf("failed to set DeletionPolicy on volumesnapshotcontent %s. Skipping deletion", snapCont.Name))
+	}
+	if apierrors.IsNotFound(err) {
```
Comment (dsu-igeek) [inline internal/delete/volumesnapshotcontent_action.go:58]: I think this makes more sense to move inside the if err != nil statement and then do if IsNotFound/else inside of that.

File: internal/delete/volumesnapshotcontent_action.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		p.Log.Infof("volumesnapshotcontent %s not found", snapCont.Name)
+		return nil
+	}
+	err = snapClient.SnapshotV1beta1().VolumeSnapshotContents().Delete(context.TODO(), snapCont.Name, metav1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
```
Comment (dsu-igeek) [inline internal/delete/volumesnapshotcontent_action.go:63]: We should log if we're getting IsNotFound here.  It shouldn't happen and I agree that returning nil is the right thing, but if it is happening, something strange is going on in the system and it doesn't hurt to log it.  There may be a race with something else that we're not aware of.

File: internal/delete/volumesnapshot_action.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			return errors.Wrapf(err, fmt.Sprintf("failed to patch DeletionPolicy of volume snapshot %s/%s", vs.Namespace, vs.Name))
+		}
+
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+	}
+	err = snapClient.SnapshotV1beta1().VolumeSnapshots(vs.Namespace).Delete(context.TODO(), vs.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return err
```
Comment (dsu-igeek) [inline internal/delete/volumesnapshot_action.go:68]: The VolumeSnapshotContent is returning nil if it gets NotFound here.  I think we should have the same behavior between the two.

File: internal/util/util.go [hunk 6]
```diff
@@ -273,3 +276,19 @@ func IsVolumeSnapshotExists(volSnap *snapshotv1beta1api.VolumeSnapshot, snapshot
 
 	return exists
 }
+
+func SetVolumeSnapshotContentDeletionPolicy(vscName string, csiClient snapshotter.SnapshotV1beta1Interface) error {
+	pb := []byte(`{"spec":{"deletionPolicy":"Delete"}}`)
+	_, err := csiClient.VolumeSnapshotContents().Patch(context.TODO(), vscName, types.MergePatchType, pb, metav1.PatchOptions{})
+	if err != nil {
```
Comment (dsu-igeek) [inline internal/util/util.go:283]: This can just be return err

**Review by dsu-igeek — CHANGES_REQUESTED** [review]
Looks good, a few minor things, thanks!

File: internal/util/util.go (hunk 6 above)
Comment (ashish-amarnath) [reply internal/util/util.go]: resolved!

File: internal/delete/volumesnapshot_action.go (hunk 5 above)
Comment (ashish-amarnath) [reply internal/delete/volumesnapshot_action.go:68]: resolved!

**Review by carlisia — APPROVED** [review]
lgtm 👍

**Pull request merged**